url20b0f880a5_jgraph_legacy_jgraph5_tgz-pJ8-org_jgraph_example_HelloWorldJ8/src/com/jgraph/layout/organic/JGraphOrganicLayout.java

```diff
@@ -0,0 +1,1581 @@
+	 *            this node
+	 * @param oldAdditionalFactorsEnergy
+	 *            The previous energy cost for additional factors from
+	 *            sub-classes
+	 * 
+	 * @return the delta of the new energy cost to the old energy cost
+	 * 
+	 */
+	protected double calcEnergyDelta(int index, double oldNodeDistribution,
+			double oldEdgeDistance, double oldEdgeCrossing,
+			double oldBorderLine, double oldEdgeLength, double oldAdditionalFactorsEnergy) {
+		double energyDelta = 0.0;
+		energyDelta += getNodeDistribution(index) * 2.0;
+		energyDelta -= oldNodeDistribution * 2.0;
+
+		energyDelta += getBorderline(index);
+		energyDelta -= oldBorderLine;
+
+		energyDelta += getEdgeDistanceFromNode(index);
+		energyDelta += getEdgeDistanceAffectedNodes(index);
+		energyDelta -= oldEdgeDistance;
+
+		energyDelta -= oldEdgeLength;
+		energyDelta += getEdgeLengthAffectedEdges(index);
+		
+		energyDelta -= oldEdgeCrossing;
+		energyDelta += getEdgeCrossingAffectedEdges(index);
+
+		energyDelta -= oldAdditionalFactorsEnergy;
+		energyDelta += getAdditionFactorsEnergy(index);
+		
+		return energyDelta;
+	}
+	
+	/**
+	 * Creates a random permutation of the numbers from 0 to 
+	 * <code>length</code>
+	 * 
+	 * @param length
+	 *            number of different numbers to be generated
+	 * @return Permutation of the Numbers between 0 and <code>length</code>
+	 */
+
+	protected int[] createPermutation(int length) {
+		int[] permutation = new int[length];
+		for (int i = 0; i < permutation.length; i++)
+			permutation[i] = i;
+		for (int i = 0; i < permutation.length; i++) {
+			int t = permutation[i];
+			int randomIndex = (int) (Math.random() * (double) (length - 1));
+			// Swap values
+			permutation[i] = permutation[randomIndex];
+			permutation[randomIndex] = t;
+		}
+		return permutation;
+	}
+
+	/**
+	 * Calculates the energy cost of the specified node relative to all other
+	 * nodes. Basically produces a higher energy the closer nodes are together.
+	 * 
+	 * @param i the index of the node in the array <code>v</code>
+	 * @return the total node distribution energy of the specified node 
+	 */
+	protected double getNodeDistribution(int i) {
+		double energy = 0.0;
+		
+		// This check is placed outside of the inner loop for speed, even
+		// though the code then has to be duplicated
+		if (isOptimizeNodeDistribution == true) {
+			if (approxNodeDimensions) {
+				for (int j = 0; j < v.length; j++) {
+					if (i != j) {
+						double vx = v[i].x - v[j].x;
+						double vy = v[i].y - v[j].y;
+						double distanceSquared = vx * vx + vy * vy;
+						distanceSquared -= v[i].radiusSquared;
+						distanceSquared -= v[j].radiusSquared;
+						
+						// prevents from dividing with Zero.
+						if (distanceSquared < minDistanceLimitSquared) {
+							distanceSquared = minDistanceLimitSquared;
+						}
+						
+						energy += nodeDistributionCostFactor / distanceSquared;
+					}
+				}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 723:738, 739:759.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.